Lib/asyncio/coroutines.py

```diff
@@ -259,17 +196,14 @@ def wrapper(*args, **kwds):
 
 def iscoroutinefunction(func):
     """Return True if func is a decorated coroutine function."""
-    return (getattr(func, '_is_coroutine', None) is _is_coroutine or
-            _inspect_iscoroutinefunction(func))
+    return (inspect.iscoroutinefunction(func) or
+            getattr(func, '_is_coroutine', None) is _is_coroutine)
 
 
-_COROUTINE_TYPES = (types.GeneratorType, CoroWrapper)
-if _CoroutineABC is not None:
-    _COROUTINE_TYPES += (_CoroutineABC,)
-if _types_CoroutineType is not None:
-    # Prioritize native coroutine check to speed-up
-    # asyncio.iscoroutine.
-    _COROUTINE_TYPES = (_types_CoroutineType,) + _COROUTINE_TYPES
+# Prioritize native coroutine check to speed-up
+# asyncio.iscoroutine.
+_COROUTINE_TYPES = (types.CoroutineType, Coroutine,
```

Comment:
In the current code the natural type `types.GeneratorType` is tested before the abstract type `Coroutine`.

Reply:
When async/await is used `types.CoroutineType` and `collections.abc.Coroutine` are more frequent than old styled generators.
That's why I've changed the order.

Reply:
But checking an instance of `types.GeneratorType` is much faster than running the Python code in `collections.abc.Coroutine.__subclasshook__`.